include/platform/messagepool.h

```diff
@@ -0,0 +1,100 @@
+/*
+ *  Copyright (c) 2016, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for the message pool.
+ */
+
+#ifndef MESSAGEPOOL_H_
+#define MESSAGEPOOL_H_
+
+#include <stdint.h>
+#include <openthread-types.h>
+
+/**
+ * @defgroup messagepool MessagePool
+ * @ingroup platform
+ *
+ * @brief
+ *   This module includes the platform abstraction for the message pool.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This structure contains a pointer to the next Message buffer.
+ *
+ */
+struct BufferHeader
+{
+    struct BufferHeader *mNext;  ///< A pointer to the next Message buffer.
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Initialize the platform implemented message pool with the provided buffer list.
+ *
+ * @param[in] aFreeBufferList   A linked list of free buffers.
+ * @param[in] aNumFreeBuffers   A pointer to an int containing the number of free buffers in aFreeBufferList
+ *
+ */
+void otPlatMessagePoolInit(struct BufferHeader *aFreeBufferList, int *aNumFreeBuffers);
```

Comment:
Ok. That was really helpful. I understand what you are trying to do now. You want to be able to allocate messages on a different thread. That causes you to require your allocation to be multi-threaded.

It might still be better to go to a dynamic memory allocation model instead of going this route. OpenThread, and your TCP stack code could both call into the same dynamic allocation code, that may not need to be synchronized with each other at all (just a general malloc or calloc call). If you would like, I could look at adapting these message classes to use the otPlatCAlloc function to dynamically alloc messages.